aidb/engine/approx_select_engine.py

```diff
@@ -0,0 +1,224 @@
+from typing import List
+
+from aidb.engine.tasti_engine import TastiEngine
+from aidb.query.query import Query
+from aidb.utils.constants import VECTOR_ID_COLUMN
+
+PROXY_SCORE = 'proxy_score'
+MASS = 'mass'
+WEIGHT = 'weight'
+BUDGET = 10000
+
+class ApproxSelectEngine(TastiEngine):
+  # TODO: design a better algorithm, this function is same as the function in Limit Engine
+  def score_fn(self, score_for_all_df: pd.DataFrame, score_connected: List[List[str]]) -> pd.Series:
+    '''
+    convert query result to score, return a Dataframe contains score column, the index is blob index
+    if A and B, then the score is min(score A, score B)
+    if A or B, then the score is max(score A, score B)
+    '''
+    proxy_score_all_blobs = np.zeros(len(score_for_all_df))
+    for idx, (index, row) in enumerate(score_for_all_df.iterrows()):
+      min_score = 1
+      for or_connected in score_connected:
+        max_score = 0
+        for score_name in or_connected:
+          max_score = max(max_score, row[score_name])
+        min_score = min(min_score, max_score)
+      proxy_score_all_blobs[idx] = min_score
+    return pd.Series(proxy_score_all_blobs, index=score_for_all_df.index)
+
+
+  async def execute_approx_select_query(self, query: Query):
+    query.is_valid_approx_select_query()
+    # generate proxy score for each blob
+    score_for_all_df, score_connected = await self.get_proxy_scores_for_all_blobs(query)
+
+    proxy_score_for_all_blobs = self.score_fn(score_for_all_df, score_connected)
+
+    dataset = self.get_sampled_proxy_blob(proxy_score_for_all_blobs)
+
+    # This is used for parallel test
+    seed = (mp.current_process().pid * np.random.randint(100000, size=1)[0]) % (2**32 - 1)
+    sampled_df = dataset.sample(BUDGET, replace=True, weights=WEIGHT, random_state=seed)
+
+
+    async with self._sql_engine.begin() as conn:
+      satisfied_sampled_results, all_sampled_results = await self.get_inference_results(
+          query,
+          list(sampled_df.index),
+          conn
+      )
+
+      satisfied_sampled_results = satisfied_sampled_results.join(sampled_df, how='inner')
+      sorted_satisfied_sampled_results = satisfied_sampled_results.sort_values(by=PROXY_SCORE, ascending=False)
+
+      no_result_length = BUDGET - len(sampled_df.loc[list(set(all_sampled_results.index))])
+
+      all_sampled_results = all_sampled_results.join(sampled_df, how='inner')
+
+      total_length = no_result_length + len(all_sampled_results)
+
+      recall_target = query.recall_target
+      confidence = query.confidence / 100
+      tau_modified = self.tau_modified(
+          sorted_satisfied_sampled_results,
+          recall_target,
+          confidence,
+          total_length
+      )
+
+      R1 = sorted_satisfied_sampled_results.index
+      R2 = dataset[dataset[PROXY_SCORE] >= tau_modified].index
+
+      additional_samples = list(set(R1).union(set(R2)))
+
+      print('num_samples', len(additional_samples))
```

Comment:
Can we turn the prints into logging and make the tests print info? Can do this in another PR but raise an issue in that case